### ISSUE TITLE
Better meshing

### ISSUE THREAD

**Issue by heyx3** [heyx3/BpWorld#14]
Currently we're using the simplest, dumbest meshing technique: create one quad for each visible voxel face. This actually doesn't work too bad for the current 64^3 grid, and the obvious advantage is that it's very fast to run.

But there are some ways it can be improved:

1. Better memory usage: pre-allocate a big array to store the result in, and keep it somewhere so it can be re-used if the user requests a new voxel scene. The meshing work is currently split across threads, so use a lock-free method of writing into this array (i.e. hold a counter, interlocked-add to the counter to reserve space).
2. After finishing this first-pass mesh, spawn a new task to run in the background and work out a more efficient mesh. Once the task is finished, on the main thread replace the older/worse mesh with the newer/better one.

**Comment by heyx3** [comment]
3. Explore GPU-compute-based meshing

**Comment by heyx3** [comment]
Completed the improvements in memory usage and use of threading. The mesher splits work across threads (using `@threads`), and writes the results into a single big lock-free buffer.

Also implemented a "preview" version of voxel materials, using the geometry shader to generate voxel faces. This is significantly slower than the meshed version, but can be displayed the moment we have a 3D voxel texture, while meshing is still being computed on another thread.

GPU compute-based meshing is still an area worth investigating for larger scenes in the future.

**Comment by heyx3** [comment]
Replacing this issue with a more focused one.
